### ISSUE TITLE
[Phase 6] Evaluation Metrics & Pipeline Integration

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/real-estate-valuation-engine#8]
## Summary

Implement evaluation metrics and integrate all components into a unified end-to-end pipeline. This phase completes the system by connecting data loading, feature engineering, modeling, and evaluation.

**Part of:** #1
**Depends on:** #7 (Phase 5)

## Objectives

- [ ] Implement `MetricsCalculator` class with MAPE, MedAPE, RMSE, MAE
- [ ] Add PICP and MPIW for interval evaluation
- [ ] Implement `ValuationPipeline` class
- [ ] Add YAML configuration loading
- [ ] Integrate cross-validation into pipeline
- [ ] Write integration tests

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/evaluation/metrics.py` | 1-60 | Implement | MetricsCalculator class |
| `src/pipeline/pipeline.py` | 1-150 | Implement | ValuationPipeline class |
| `tests/pipeline/test_pipeline.py` | NEW | Create | Integration tests |

### MetricsCalculator

```python
# src/evaluation/metrics.py
import numpy as np
from dataclasses import dataclass

@dataclass
class EvaluationMetrics:
    """Container for all evaluation metrics."""
    mape: float          # Mean Absolute Percentage Error
    medape: float        # Median APE (robust)
    rmse: float          # Root Mean Squared Error
    mae: float           # Mean Absolute Error
    picp: float = None   # Prediction Interval Coverage
    mpiw: float = None   # Mean Prediction Interval Width

class MetricsCalculator:
    """Calculate valuation accuracy metrics."""

    def calculate(self, y_true, y_pred, y_lower=None, y_upper=None):
        # Point prediction metrics
        ape = np.abs(y_true - y_pred) / y_true
        mape = np.mean(ape) * 100
        medape = np.median(ape) * 100
        rmse = np.sqrt(np.mean((y_true - y_pred) ** 2))
        mae = np.mean(np.abs(y_true - y_pred))

        # Interval metrics
        picp, mpiw = None, None
        if y_lower is not None and y_upper is not None:
            in_interval = (y_true >= y_lower) & (y_true <= y_upper)
            picp = np.mean(in_interval)
            mpiw = np.mean((y_upper - y_lower) / y_pred) * 100

        return EvaluationMetrics(
            mape=mape, medape=medape, rmse=rmse, mae=mae,
            picp=picp, mpiw=mpiw
        )
```

### ValuationPipeline

```python
# src/pipeline/pipeline.py
import yaml
import pandas as pd
import geopandas as gpd
from pathlib import Path

from src.data.loader import DataLoader
from src.data.imputer import MissingValueHandler
from src.features.spatial import SpatialFeatureEngine
from src.features.temporal import TemporalFeatureEngine
from src.models.ensemble import EnsembleValuationModel
from src.evaluation.cross_validation import TemporalCrossValidator
from src.evaluation.metrics import MetricsCalculator

class ValuationPipeline:
    """End-to-end pipeline for property valuation."""

    def __init__(self, config_path=None):
        self.config = self._load_config(config_path)
        self.loader = DataLoader()
        self.imputer = None
        self.spatial_engine = None
        self.temporal_engine = None
        self.model = None
        self.cv = TemporalCrossValidator()
        self.metrics = MetricsCalculator()
        self._raw_data = None

    def _load_config(self, path):
        if path is None:
            return {}
        with open(path) as f:
            return yaml.safe_load(f)

    def load_data(self, filepath):
        self._raw_data = self.loader.load_csv(filepath)
        return self

    def fit(self):
        # Initialize components
        self.imputer = MissingValueHandler(
            numeric_features=['sqft', 'lot_size', 'bedrooms', 'bathrooms', 'year_built'],
            categorical_features=['property_type']
        )
        self.spatial_engine = SpatialFeatureEngine()
        self.temporal_engine = TemporalFeatureEngine()
        self.model = EnsembleValuationModel()

        # Process data
        data_imputed = self.imputer.fit_transform(self._raw_data)
        spatial_features = self.spatial_engine.fit_transform(self._raw_data)
        temporal_features = self.temporal_engine.transform(data_imputed)

        # Combine features
        property_features = data_imputed[['sqft', 'lot_size', 'bedrooms', 'bathrooms', 'year_built']]
        features = pd.concat([property_features, spatial_features, temporal_features], axis=1)

        # Fit model
        self.model.fit(features, self._raw_data['sale_price'])
        return self

    def predict(self, new_data, return_intervals=True):
        # Process new data using fitted transformers
        data_imputed = self.imputer.transform(new_data)
        spatial_features = self.spatial_engine.transform(new_data)
        temporal_features = self.temporal_engine.transform(data_imputed)

        property_features = data_imputed[['sqft', 'lot_size', 'bedrooms', 'bathrooms', 'year_built']]
        features = pd.concat([property_features, spatial_features, temporal_features], axis=1)

        result = self.model.predict(features)

        output = pd.DataFrame({'predicted_value': result.point_estimate}, index=new_data.index)
        if return_intervals and result.lower_bound is not None:
            output['lower_bound'] = result.lower_bound
            output['upper_bound'] = result.upper_bound

        return output
```

### Integration Test

```python
# tests/pipeline/test_pipeline.py
import pytest
from src.pipeline import ValuationPipeline

def test_pipeline_end_to_end(fixtures_path):
    """Pipeline should run end-to-end on sample data."""
    pipeline = ValuationPipeline(config_path="configs/default.yaml")
    pipeline.load_data(fixtures_path / "sample_properties.csv")
    pipeline.fit()

    # Predict on same data (smoke test)
    predictions = pipeline.predict(pipeline._raw_data)

    assert 'predicted_value' in predictions.columns
    assert len(predictions) == len(pipeline._raw_data)
    assert predictions['predicted_value'].notna().all()
```

## Verification

```bash
# Run integration tests
pytest tests/pipeline/test_pipeline.py -v

# End-to-end test
python -c "
from src.pipeline import ValuationPipeline

pipeline = ValuationPipeline(config_path='configs/default.yaml')
pipeline.load_data('tests/fixtures/sample_properties.csv')
pipeline.fit()

predictions = pipeline.predict(pipeline._raw_data)
print(predictions.head())
print(f'\nMean predicted value: ${predictions[\"predicted_value\"].mean():,.0f}')
"
```

## Definition of Done

- [ ] Pipeline runs end-to-end without errors
- [ ] Predictions have reasonable values
- [ ] Configuration file controls behavior
- [ ] Cross-validation produces metrics
- [ ] Integration tests pass
- [ ] MAPE metric calculated correctly

## Success Metrics Target

| Metric | Target |
|--------|--------|
| MAPE | < 15% |
| PICP | > 85% |
| No data leakage | Verified |

## Labels

`phase-6`, `pipeline`, `integration`, `enhancement`
